### ISSUE TITLE
Issues with `PYTHONPATH` resolution in recent python/rules_python versions

### ISSUE THREAD

**Issue by phlax** [bazelbuild/rules_python#1221]
# 🐞 bug report

### Affected Rule

 The issue is caused by the rule: 

afaict ~any python rules but ive tested or hit the issue with `py_binary`, `py_library`

### Is this a regression?

 Yes, the previous version in which this bug was not present was: ....

We had trouble upgrading to 0.19.x versions iirc due to this issue

We have been able to upgrade to 0.20.0 but when upgrading the python version to 3.11 the issue reappears

### Description

 A clear and concise description of the problem...

Expected paths are not being added to `sys.path`

## 🔬 Minimal Reproduction

```starlark
py_library(
    name = "mylib",
    srcs = ["mylib.py"],
)

py_binary(
    name = "mybin",
    srcs = ["mybin.py"],
    deps = [":mylib"],
)

```

```python

def myfun():
    print("BOOM!")

```

```python
from mylib import myfun


def main():
    myfun()


if __name__ == "__main__":
    main()

```

## 🔥 Exception or Error

```console
$ bazel run //tools/base:mybin
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 19b6db42-4a15-4858-be40-7cf804c798de
INFO: Analyzed target //tools/base:mybin (126 packages loaded, 3458 targets configured).
INFO: Found 1 target...
Target //tools/base:mybin up-to-date:
  bazel-bin/tools/base/mybin
INFO: Elapsed time: 4.511s, Critical Path: 0.28s
INFO: 4 processes: 4 internal.
INFO: Build completed successfully, 4 total actions
INFO: Running command line: bazel-bin/tools/base/mybin
Traceback (most recent call last):
  File "/home/worker/.cache/bazel/_bazel_worker/f704bab1b165ed1368cb88f9f49e7532/execroot/envoy/bazel-out/k8-fastbuild/bin/tools/base/mybin.runfiles/envoy/tools/base/mybin.py", line 2, in <module>
    from mylib import myfun
ModuleNotFoundError: No module named 'mylib'

```

## 🌍  Your Environment

**Operating System:**

<pre>
  <code>
Linux
  </code>
</pre>

**Output of `bazel version`:**

<pre>
  <code>
$ bazel version
Bazelisk version: v1.11.0
Build label: 6.1.0
Build target: bazel-out/k8-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Mon Mar 6 17:09:47 2023 (1678122587)
Build timestamp: 1678122587
Build timestamp as int: 1678122587

  </code>
</pre>

**Rules_python version:**

<pre>
  <code>
0.20.0
  </code>
</pre>

**Anything else relevant?**

Python version: 3.11

**Comment by phlax** [comment]
cc @htuch

**Comment by phlax** [comment]
adding a shim - ie:

```python
import os
import sys

sys.path += [os.path.dirname(__file__)]

```

works, and is what i have done to workaround, either in our py targets, or for py files in dep rules

**Comment by phlax** [comment]
> Can you elaborate on how you hooked that shim in?

very hackily - see this PR for details https://github.com/envoyproxy/envoy/pull/25359/files

seems to have survived a `rules_python` upgrade, i didnt check to see if the hacks could be removed yet

**Comment by alexeagle** [comment]
I expect this is resolved by https://github.com/aspect-build/rules_py because it doesn't mess with the PYTHONPATH, it just lays out a python-idiomatic virtualenv with a `site-packages` folder.

**Comment by phlax** [comment]
interesting we just landed a workaround to inject namespacing init pys to our repo folders to avoid pypi namespace clashing - seems like we could avoid that workaround - ill look further when i get chance

**Comment by phlax** [comment]
this is still an issue

**Comment by alexeagle** [comment]
Note, I deleted the stale bot :)

**Comment by aignas** [comment]
Well, this is actually working as intended. the `py_library` is missing an `imports` attribute in the reproduction, so if I modify the example and run it, it works:
```
$ bazel run //tools/base:mybin
INFO: Analyzed target //tools/base:mybin (1 packages loaded, 4 targets configured).
INFO: Found 1 target...
Target //tools/base:mybin up-to-date:
  bazel-bin/tools/base/mybin
INFO: Elapsed time: 1.004s, Critical Path: 0.05s
INFO: 2 processes: 3 action cache hit, 2 internal.
INFO: Build completed successfully, 2 total actions
INFO: Running command line: bazel-bin/tools/base/mybin
BOOM!

$ git diff
diff --git a/tools/base/BUILD.bazel b/tools/base/BUILD.bazel
index 5c40070..63324c4 100644
--- a/tools/base/BUILD.bazel
+++ b/tools/base/BUILD.bazel
@@ -4,6 +4,7 @@ load("@rules_python//python:py_library.bzl", "py_library")
 py_library(
     name = "mylib",
     srcs = ["mylib.py"],
+    imports = ["."],
 )

 py_binary(
```

So I'll close as won't do. The previous behaviour might have been due to `3.10` allowing the interpreter to escape sandbox and with `3.11` there was an extra `safe_path` feature added that the entrypoint template in the `bazel` repository (this is the version when the issue was raised) was using.

The correct way to fix this would be to either wait until #2156 is resolved or add `imports` attributes to your `py_library` instances.
